### PR TITLE
config/v1/types_oauth: add OAuthAudit

### PR DESCRIPTION
**Pull request by s-urbaniak**
This adds audit configuration to oauth config type.

/cc @stlaz @EmilyM1 @ibihim

### CODE REVIEW

File: config/v1/types_oauth.go [hunk 1]
```diff
@@ -40,8 +40,27 @@ type OAuthSpec struct {
 	// templates allow you to customize pages like the login page.
 	// +optional
 	Templates OAuthTemplates `json:"templates"`
+	// Audit specifies what needs to be audited.
+	// + optional
+	Audit OAuthAudit `json:"audit"`
 }
 
+type OAuthAudit struct {
```
Comment (stlaz) [inline config/v1/types_oauth.go:48]: should have godoc

File: config/v1/types_oauth.go [hunk 2]
```diff
@@ -40,8 +40,27 @@ type OAuthSpec struct {
 	// templates allow you to customize pages like the login page.
 	// +optional
 	Templates OAuthTemplates `json:"templates"`
+	// Audit specifies what needs to be audited.
```
Comment (stlaz) [inline config/v1/types_oauth.go:43]: this could probably use a better description

File: config/v1/types_oauth.go [hunk 3]
```diff
@@ -40,8 +40,27 @@ type OAuthSpec struct {
...
+	// + optional
+	Audit OAuthAudit `json:"audit"`
 }
 
+type OAuthAudit struct {
+	// Simple drop in profile type.
+	Profile OAuthAuditProfileType `json:"profile,omitempty"`
+}
+
+type OAuthAuditProfileType string
```
Comment (stlaz) [inline config/v1/types_oauth.go:53]: should have godoc

File: config/v1/types_oauth.go [hunk 4]
```diff
@@ -40,8 +40,27 @@ type OAuthSpec struct {
...
 	// +optional
 	Templates OAuthTemplates `json:"templates"`
+	// Audit specifies what needs to be audited.
+	// + optional
+	Audit OAuthAudit `json:"audit"`
 }
 
+type OAuthAudit struct {
+	// Simple drop in profile type.
+	Profile OAuthAuditProfileType `json:"profile,omitempty"`
```
Comment (ibihim) [inline config/v1/types_oauth.go:50]: We would need to set here a default value here with the kubebuilder, right?

https://github.com/openshift/enhancements/pull/894/files#r721093860

File: config/v1/types_oauth.go [hunk 5]
```diff
@@ -40,8 +40,31 @@ type OAuthSpec struct {
...
 	Templates OAuthTemplates `json:"templates"`
+	// audit specifies what should be audited in the context of OAuthServer.
+	// +optional
+	Audit OAuthAudit `json:"audit"`
+}
+
+// OAuthAudit specifies the Audit profile in use.
+type OAuthAudit struct {
+	// profile is a simple drop in profile type.
+	// +kubebuilder:default:="WriteLoginEvents"
```
Comment (stlaz) [inline config/v1/types_oauth.go:51]: I wonder how this works if the profile is unspecified. What would happen if we add the `+optional` tag here?

File: config/v1/types_oauth.go [hunk 6]
```diff
@@ -40,8 +40,31 @@ type OAuthSpec struct {
...
 	// +optional
 	Templates OAuthTemplates `json:"templates"`
+	// audit specifies what should be audited in the context of OAuthServer.
+	// +optional
+	Audit OAuthAudit `json:"audit"`
+}
+
+// OAuthAudit specifies the Audit profile in use.
+type OAuthAudit struct {
+	// profile is a simple drop in profile type.
```
Comment (stlaz) [inline config/v1/types_oauth.go:50]: Mention possible values?

File: config/v1/types_oauth.go [hunk 7]
```diff
@@ -40,8 +40,31 @@ type OAuthSpec struct {
...
+	// +kubebuilder:default:="WriteLoginEvents"
+	Profile OAuthAuditProfileType `json:"profile,omitempty"`
 }
 
+// OAuthAuditProfileType is an enum that defines audit profiles.
```
Comment (stlaz) [inline config/v1/types_oauth.go:55]: A common user does not know what an enum is.

`OAuthAuditProfileType specifies high-level configuration of the oauth-server audit profiles` maybe?

File: config/v1/types_oauth.go (hunk 4 above)
Comment (stlaz) [reply config/v1/types_oauth.go:52]: yes

File: config/v1/types_oauth.go (hunk 5 above)
Comment (ibihim) [reply config/v1/types_oauth.go:51]: Ok, so should we maybe delete the optional flag from [Audit](https://github.com/openshift/api/pull/1042/files#diff-17be816abc6f9bcc605991fa9b3aebe956afbb04c48c0ff957216993f3728438R45)? If it is optional, the probability that the default value will be autogenerated is minimal.

File: config/v1/types_oauth.go (hunk 7 above)
Comment (ibihim) [reply config/v1/types_oauth.go:55]: Oh, you mean non-developers read that? Good to know the audience! thx

File: config/v1/types_oauth.go (hunk 6 above)
Comment (ibihim) [reply config/v1/types_oauth.go:50]: I guess this is another audience thing :smile:

File: config/v1/types_oauth.go (hunk 5 above)
Comment (ibihim) [reply config/v1/types_oauth.go:52]: @sttts, could you verify that the thought process of removing `+optional` is right and tell @stlaz and me what would happen on an `+optional`?

File: config/v1/types_oauth.go [hunk 8]
```diff
@@ -40,8 +40,33 @@ type OAuthSpec struct {
 	// templates allow you to customize pages like the login page.
 	// +optional
 	Templates OAuthTemplates `json:"templates"`
+	// audit specifies what should be audited in the context of OAuthServer.
```
Comment (stlaz) [inline config/v1/types_oauth.go:43]: what happens after an upgrade when this is not optional?

**Review by stlaz — CHANGES_REQUESTED** [review]

File: config/v1/0000_10_config-operator_01_oauth.crd.yaml [hunk 9]
```diff
@@ -36,6 +36,19 @@ spec:
...
+                  default:
+                    profile: WriteLoginEvents
+                  properties:
+                    profile:
+                      description: profile is a simple drop in profile type that can be turned off by setting it to "None" or it can be turned on by setting it to "WriteLoginEvents". By default the profile is set to "WriteLoginEvents".
+                      type: string
+                      default: WriteLoginEvents
+                      enum:
+                        - None
+                        - WriteLoginEvents
```
Comment (sttts) [inline config/v1/0000_10_config-operator_01_oauth.crd.yaml:51]: sgtm

File: config/v1/0000_10_config-operator_01_oauth.crd.yaml [hunk 10]
```diff
@@ -36,6 +36,19 @@ spec:
               description: spec holds user settable values for configuration
               type: object
               properties:
+                audit:
+                  description: audit specifies what should be audited in the context of OAuthServer. By default the Audit is turned on.
```
Comment (sttts) [inline config/v1/0000_10_config-operator_01_oauth.crd.yaml:40]: nit (follow-up is fine): "the Audit" is what?

**Pull request merged**